bpo-30860: Consolidate stateful runtime globals.

ericsnowcurrently opened the pull request:
1. group the (stateful) runtime globals into various topical structs
2. consolidate the topical structs under a single top-level _PyRuntimeState struct
3. add a check-c-globals.py script that helps identify runtime globals

Other globals are excluded (see globals.txt and check-c-globals.py).

<!-- issue-number: bpo-30860 -->
https://bugs.python.org/issue30860
<!-- /issue-number -->

ericsnowcurrently committed:
Add the check-c-globals script.

ericsnowcurrently committed:
Clean up the check-c-globals script.

ericsnowcurrently committed:
Update the list of "runtime" globals.

ericsnowcurrently committed:
Handle builtin modules via a pattern check.

ericsnowcurrently committed:
Rely on a pattern check for static type-related vars.

ericsnowcurrently committed:
Handle most exceptions and warnings via pattern checks.

ericsnowcurrently committed:
Clean up _is_type_var().

ericsnowcurrently committed:
Add runtime vars for builtin modules.

ericsnowcurrently committed:
Ignore trailing comments in globals-runtime.txt.

ericsnowcurrently committed:
Add runtime vars for builtin types.

ericsnowcurrently committed:
Add runtime vars for compiler.

ericsnowcurrently committed:
Add runtime vars for interpreter.

ericsnowcurrently committed:
Add runtime vars for import.

ericsnowcurrently committed:
Add runtime vars for startup.

ericsnowcurrently committed:
Add other runtime vars.

ericsnowcurrently committed:
Separate the 3 runtime global "scopes".

ericsnowcurrently committed:
Do not worry about a possible race.

ericsnowcurrently committed:
Narrow down the globals lists.

ericsnowcurrently committed:
runtime -> core.

ericsnowcurrently committed:
Hold up more globals.

ericsnowcurrently committed:
Add _PyRuntimeState and _PyRuntime.

ericsnowcurrently committed:
Pull pylifecycle globals into _PyRuntimeState.

ericsnowcurrently committed:
Pull mem globals into _PyRuntimeState.

ericsnowcurrently committed:
Pull warnings globals into _PyRuntimeState.

ericsnowcurrently committed:
Move nb_threads into PyInterpreterState.

ericsnowcurrently committed:
Pull ceval globals into _PyRuntimeState.

ericsnowcurrently committed:
Pull gil into _PyRuntimeState.

ericsnowcurrently committed:
Move check_interval, warnoptions, and xoptions into PyInterpreterState.

ericsnowcurrently committed:
Move pythread_stacksize into PyInterpreterState.

ericsnowcurrently committed:
Pull the rest of the GIL globals into _PyRuntime.

ericsnowcurrently committed:
Factor out _gil_globals and _gilstate_globals structs.

ericsnowcurrently committed:
Factor out globals structs.

ericsnowcurrently committed:
Factor out a _pending_calls struct.

ericsnowcurrently committed:
Update globals list.

ericsnowcurrently committed:
Ignore a few more globals.

ericsnowcurrently committed:
Only track one list of globals.

ericsnowcurrently committed:
Mark _Py_HashSecret* as global.

ericsnowcurrently committed:
_Py_TracingPossible -> _PyRuntime.ceval.tracing_possible.

ericsnowcurrently committed:
Re-structure globals.txt.

ericsnowcurrently committed:
Ignore the symtable globals.

ericsnowcurrently committed:
_Py_CoreInitialized -> _PyRuntime.core_initialized.

ericsnowcurrently committed:
Pull in the list of interpreters.

ericsnowcurrently committed:
Clean up globals.txt.

ericsnowcurrently committed:
Ignore the import lock.

ericsnowcurrently committed:
Factor out pymalloc.h.

ericsnowcurrently committed:
Fix func header formatting.

ericsnowcurrently committed:
Move "globals" structs to the appropriate header files.

ericsnowcurrently committed:
Pull usedpools into _PyRuntime.

ericsnowcurrently committed:
Drop a dead comment.

ericsnowcurrently committed:
Pull in gcmodule.c.

ericsnowcurrently committed:
Clean up globals.txt.

ericsnowcurrently committed:
Pass pointers to the *_Initialize() funcs.

1st1 commented:
An obvious nit: there are a lot of lines that are longer than 79 chars. Those will need to be fixed per PEP 7.

ericsnowcurrently committed:
Clean up line length.

ericsnowcurrently committed:
Add a filename CLI arg.

ericsnowcurrently committed:
Move check-c-globals into Tools/c-globals.

ericsnowcurrently committed:
Add a README.

ericsnowcurrently committed:
Fix typo in header guard.

ericsnowcurrently commented:
FYI, there are test failures in test_capi, test_warnings, and test_tracemalloc.  I'll address those tomorrow, but I expect the fixes won't require many changes.

ncoghlan commented on Tools/c-globals/README:
```diff
@@ -0,0 +1,40 @@
+* module state
+* Python runtime state
+
+The globals.txt file is organized similarly.  Of the the different
```
`the the`

ericsnowcurrently replied:
fixed

ncoghlan commented on Tools/c-globals/README:
```diff
@@ -0,0 +1,40 @@
+
+The globals.txt file is organized similarly.  Of the the different
+categories, the last two are problematic and generally should not exist
+in the codebase.  Globals that hold module state (i.e. in Modules/*.c)
```
Paragraph break here, as it isn't clear that you've switched from talking about "the last two" points to just the second-last one.

ericsnowcurrently replied:
fixed

ncoghlan commented on Include/pystate.h:
```diff
@@ -185,6 +204,74 @@ typedef struct _ts {
+
+    struct pyinterpreters {
+#ifdef WITH_THREAD
+#ifdef Py_BUILD_CORE
```
Having these `Py_BUILD_CORE` conditionals inside the struct definitions doesn't look right to me - don't they mean that the compiler will generate incorrect field offsets when compiled *without* `Py_BUILD_CORE` set?

Even having a `Py_BUILD_CORE` block at the *end* of the struct definitions seems problematic, since we later compose them all into a single struct, where the non-core builds would again get the field sizes wrong.

So if we're going to have a distinction between "public" runtime state (accessible via macros in the public C API) and "private" runtime state (accessible only when "Py_BUILD_CORE" is set), it seems to me that distinction is going to need to be handled by having two different top level runtime structs (one public and one private), and having a similar split for every section that includes non-public process global state.

ericsnowcurrently replied:
Yeah, between your comment here and Antoine's concerns, I'll have to come up with something better.  I'd really like to keep a single _PyRuntimeState though.

ncoghlan replied:
You *might* be able to manage a single flat struct by putting the private state at the end, but honestly, I think you'll have an easier time of it by starting with two structs (_PyRuntimeState and _PyRuntimePublicState), and then progressively reducing the public state to be as close to empty as possible, and instead requiring indirection through accessor and mutator functions.

ncoghlan commented on Tools/c-globals/README:
```diff
@@ -0,0 +1,40 @@
+
+If it reports any globals then they should be resolved.  If the globals
+are runtime state then they should be folded into _PyRuntimeState.
+Otherwise they should be added to globals.txt.
```
Perhaps rename that file `ignored-globals.txt`?

ericsnowcurrently replied:
done

ncoghlan reviewed (commented):
I still like the general idea, but the way the struct definitions are currently handling `Py_BUILD_CORE` doesn't look right to me.

ericsnowcurrently committed:
Ignore _TARGET_LOCALES.

ericsnowcurrently committed:
Factor out _PyRuntimeState_Initialize().

ericsnowcurrently committed:
Only initialize _PyRuntime once.

_PyRuntime_Initialize() is equivalent to the static initialization
of all the former globals.  Since that initialization only happens
once in the process, we likewise only initialize _PyRuntime once.
However, this is not an ideal long-term solution.  _PyRuntime
initialization should be repeated at every Py_Initialze() call.  The
fact that doing so breaks the runtime reveals runtime state that is
never properly finalized.

ericsnowcurrently committed:
Do not assume that there is an active threadstate.

ericsnowcurrently committed:
Always use the global allocators.

ericsnowcurrently committed:
Factor out _allocators_globals.

ericsnowcurrently committed:
Touch up the c-globals README.

ericsnowcurrently committed:
globals.txt -> ignored-globals.txt.

ericsnowcurrently committed:
Move runtime state into "internal" header files.

ericsnowcurrently committed:
globals -> state

ericsnowcurrently committed:
Define _PyRuntime for pgen.

ericsnowcurrently committed:
Drop a superfluous include.

ericsnowcurrently committed:
Add the new header files to the Windows config.

ericsnowcurrently commented:
@zooba, any thoughts on the Windows build failure here?  I'm sure it's something trivial.

ericsnowcurrently committed:
Move _PyInterpreterState_Init() to the internal header file.

ericsnowcurrently committed:
Add _PyRuntimeState_Fini().

ericsnowcurrently committed:
Manage the interpreters mutex globally.

zooba commented:
@ericsnowcurrently Not sure that `uint` is a standard C type - I thought they were `uint#_t`, where `#` is the bit length. `unsigned int` or `uint32_t` should be okay.

ericsnowcurrently committed:
Fix Windows build.

zooba commented on Include/internal/_mem.h:
```diff
@@ -11,6 +11,9 @@ extern "C" {
 #endif
 
+#undef  uint
+#define uint    unsigned int    /* assuming >= 16 bits */
```
Oh come on, this is lazy :P It's okay for testing, but I'll insist you change all uses of `uint` into a standard type name before merging.

ericsnowcurrently replied:
It's the original code that I forgot to copy over. :)  I don't mind fixing it up though.

ericsnowcurrently committed:
"uint" -> unsigned int.

ericsnowcurrently committed:
Set up allocators in _Py_InitializeCore().

ericsnowcurrently commented:
FYI, the _Py_CheckRecursionLimit change causes a performance regression on the pickle_list and pickle_dict benchmarks.  When I get a chance I'll look into it.

ericsnowcurrently committed:
Add a NEWS entry.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

ericsnowcurrently committed:
Fix the news entry.

gpshead commented on Include/internal/_Python.h:
```diff
@@ -0,0 +1,16 @@
+/* Include all internal Python header files */
+
+#ifndef Py_BUILD_CORE
+#error "Internal headers are not available externally."
```
as Py_BUILD_CORE is for internal use, perhaps give it a _ prefix.  _Py_BUILD_CORE.  Also, do we want #define's to have lower case?  _PY_BUILD_CORE vs _Py_BUILD_CORE?

ericsnowcurrently replied:
Py_BUILD_CORE is pre-existing.  I don't mind changing it though if you still think I should.

gpshead reviewed (approved):
overall, awesome refactoring and great analysis of the global state!
